crates/ruma-api-macros/src/api/metadata.rs

```diff
@@ -160,26 +170,50 @@ impl Parse for Metadata {
         }
 
         if let Some(added) = &added {
-            if stable_path.is_none() {
+            if stable_or_r0.is_none() {
                 return Err(syn::Error::new_spanned(
                     added,
-                    "added version is defined, but no stable path exists",
+                    "added version is defined, but no stable or r0 path exists",
+                ));
+            }
+        }
+
+        if let Some(r0) = &r0_path {
+            let added = added.as_ref().expect("we error if r0 or stable is defined without added");
+
+            if added.major == 1.try_into().unwrap() && added.minor > 0 {
+                return Err(syn::Error::new_spanned(
+                    r0,
+                    "r0 defined while added version is newer than v1.0",
+                ));
+            }
+
+            if stable_path.is_none() {
+                return Err(syn::Error::new_spanned(r0, "r0 defined without stable path"));
+            }
+
+            if !r0.0.value().contains("/r0/") {
+                return Err(syn::Error::new_spanned(r0, "r0 endpoint does not contain /r0/"));
+            }
+        }
+
+        if let Some(stable) = &stable_path {
+            if stable.0.value().contains("/r0/") {
+                return Err(syn::Error::new_spanned(
+                    stable,
+                    "stable endpoint contains /r0/ (did you make a copy-paste error?)",
                 ));
             }
         }
 
         if unstable_path.is_none() && r0_path.is_none() && stable_path.is_none() {
-            // TODO replace with error
-            // return Err(syn::Error::new_spanned(metadata_kw, "no path is defined"));
-            r0_path = path.clone();
-            unstable_path = path.clone();
+            return Err(syn::Error::new_spanned(metadata_kw, "no path is defined"));
```

Comment:
I would suggest adding a little more context, for example

```suggestion
            return Err(syn::Error::new_spanned(metadata_kw, "need to define one of [r0_path, stable_path, unstable_path]"));
```